rmf_task/include/rmf_task/Request.hpp

```diff
@@ -39,6 +39,9 @@ class Request
   /// Get the id of the task
   virtual std::string id() const = 0;
 
+  /// Get the priority of the task. High priority tasks should return true.
+  virtual bool priority() const = 0;
```

Comment:
I might recommend refactoring `Request` to be a concrete class instead of a pure abstract interface.

We could have `id`, `earliest_start_time`, and `priority` be concrete fields. `id` and `earliest_start_time` would keep their current value types (`std::string` and `rmf_traffic::Time`) while `priority` could be some kind of `std::unique_ptr<Priority>` or `std::shared_ptr<Priority>` field where `Priority` is a pure abstract interface that gets generated based on whatever the "prioritization scheme" is for the cost function.

Meanwhile we could refactor the cost calculations for the planner into its own abstract interface. Different prioritization schemes can be chosen by selecting the cost calculation implementation. That cost calculation implementation would be responsible for providing the `Priority` objects that get held by the `Request` objects. For cost calculators that don't have any prioritization schemes, the `priority` field can just be left as a `nullptr`.

Reply:
I've managed to refactor `Request` as suggested but am having trouble with the cost calculation abstraction and implementation. The calculation of [assignment cost](https://github.com/osrf/rmf_core/blob/b50b5cd16eac925fa107b7c4b9748a31aa696005/rmf_task/src/rmf_task/agv/TaskPlanner.cpp#L867-L878) and [heuristic cost](https://github.com/osrf/rmf_core/blob/b50b5cd16eac925fa107b7c4b9748a31aa696005/rmf_task/src/rmf_task/agv/TaskPlanner.cpp#L880-L931) (and therefore total cost) depend heavily on the internal [Node definition](https://github.com/osrf/rmf_core/blob/b50b5cd16eac925fa107b7c4b9748a31aa696005/rmf_task/src/rmf_task/agv/TaskPlanner.cpp#L431-L495). Is there a way to proceed without making the `Node` definition part of the public API?

I considered templating the `CostCalculator` class as below. But then the sample implementation of this class (`SimpleCostCalculator`) would still need to define `Node` so that downstream packages can instantiate an object of `SimpleCostCalculator`.

```
template<typename Node>
class CostCalculator
{
public:

  /// Compute the total cost while factoring in the prioritization scheme for a node
  virtual double compute_cost(const Node& n) const = 0;

  /// Generate a priority object based on the prioritization scheme of the CostCalculator
  virtual ConstPriorityPtr priority(std::size_t value) const = 0;

  virtual ~CostCalculator() = default;
};
```
Please let me know if i'm misunderstanding something.

Reply:
That's a great point which I hadn't completely thought through.

I'd say for now let's keep the cost calculation abstraction internal so we don't have to commit the `Node` definitions to the public API. That might look something like this:

```
// Forward declaration of these abstract interfaces in a public header.
// The definitions of these abstract interfaces will remain internal.
class CostCalculator;
class Priority;

/// This is for a binary prioritization scheme. Tasks are assigned either high priority or low priority.
class HighLowPriority
{
public:

  /// Use these to assign the task priority
  static std::shared_ptr<Priority> make_low_priority();
  static std::shared_ptr<Priority> make_high_priority();

  /// Use this to give the appropriate cost calculator to the planner
  static std::shared_ptr<CostCalculator> make_cost_calculator();
};
```

The classes that implement the abstract interfaces `CostCalculator` and `Priority` can remain internal implementation details.

The one weakness I see with the approach that I'm suggesting is it can't stop users from mixing together `Priority` implementations with incompatible `CostCalculator` implementations. It may be worth thinking about how to make the API more narrow so that users can't do the wrong thing. I think the approach of forward declaring the abstract interfaces will still be useful for that, though.

Reply:
Thanks for the advice- this is a really neat approach.

Reply:
Maybe `class BinaryPriority` would be a better name than `class HighLowPriority`.

Reply:
I have refactored the package as suggested.
I renamed the above class to `BinaryPriorityScheme`. `Priority` and `CostCalculator` abstract classes are now defined as internal implementation details along with their respective sample implementations `BinaryPriority` and `BinaryPriorityCostCalculator`.
If things look okay I can proceed with updating the fleet adapter side of things.

I plan to introduce a new `Priority.msg` in `rmf_task_msgs` with a single field `uint64 value`. Then the [TaskProfile](https://github.com/osrf/rmf_core/blob/master/rmf_task_msgs/msg/TaskProfile.msg) will be updated with a new `Priority priority` field. I'm hoping having a single `value` field will be able to satisfy multiple priority schemes. Please let me know if this makes sense.